enterprise/backend/src/metabase_enterprise/query_field_validation/api.clj

```diff
@@ -0,0 +1,76 @@
+(ns metabase-enterprise.query-field-validation.api
+  (:require
+   [compojure.core :refer [GET]]
+   [metabase.api.common :as api]
+   [metabase.api.routes.common :refer [+auth]]
+   [metabase.server.middleware.offset-paging :as mw.offset-paging]
+   [toucan2.core :as t2]))
+
+(def ^:private valid-sort-columns "Columns that the card errors can be sorted by"
+  #{"name" "collection" "created_by" "last_edited_at"}) ;; We don't support sorting by error message for now
+
+(def ^:private default-sort-column "name")
+
+(def ^:private valid-sort-directions
+  #{"asc" "desc"})
+
+(def ^:private default-sort-direction "asc")
+
+(def ^:private sort-column->keyfn
+  {"name"           :name
+   "collection"     (comp :name :collection)
+   "created_by"     (comp :common_name :creator)
+   "last_edited_at" :updated_at})
+
+(defn- cards-with-inactive-fields
+  [sort-column sort-direction]
+  (let [card-id->query-fields (group-by :card_id
+                                        (t2/select :model/QueryField
+                                                   {:select [:qf.* [:f.name :column_name] [:t.name :table_name]]
+                                                    :from   [[(t2/table-name :model/QueryField) :qf]]
+                                                    :join   [[(t2/table-name :model/Card)  :c] [:= :c.id :qf.card_id]
+                                                             [(t2/table-name :model/Field) :f] [:= :f.id :qf.field_id]
+                                                             [(t2/table-name :model/Table) :t] [:= :t.id :f.table_id]]
+                                                    :where  [:and
+                                                             [:= :c.archived false]
+                                                             [:= :f.active false]
+                                                             [:= :qf.explicit_reference true]]}))
+        card-ids              (keys card-id->query-fields)
+        add-errors            (fn [{:keys [id] :as card}]
+                                (update-in card
+                                           [:errors :inactive-fields]
+                                           concat
+                                           (for [{:keys [table_name column_name]} (card-id->query-fields id)]
+                                             {:table table_name :field column_name})))]
+    ;; In theory this second query is unnecessary since we could include Card.* in the SELECT above and do more
+    ;; processing in Clojure, but since that increases the Clojure-side processing I'm not convinced it's actually any
```

Comment:
Not being able to do the sorting and pagination in the database is a huge deal. If we're going this route I think we need to also put a conservative limit in the query, and include a warning in the response if that means we're truncating the results. We can't afford to ship more non-scalable code to large customers IMHO.